include/pybind11/pybind11.h

```diff
@@ -114,9 +114,14 @@ class cpp_function : public function {
     object name() const { return attr("__name__"); }
 
 protected:
+    using unique_function_record = std::unique_ptr<detail::function_record, void (*)(detail::function_record *)>;
```

Comment:
A function pointer has state. A non-capturing function object that just calls the deleter function has no state. The standard library has a way of not storing a stateless deleter which makes `sizeof(T*) == sizeof(std::unique_ptr<T>)`. Can we make the deleter stateless?

Reading a bit more into this, I guess storing the function pointer was just convenient, because lambdas in non-evaluated context are a C++20 thing. We'll have to fall back to C++98 function objects.

Reply:
Done, I think.